### PR TITLE
perf(vm): replace snprintf with hand-rolled itoa and bounded strcat

### LINKED ISSUE #263: Replace snprintf in VM with hand-rolled itoa and strcat (drops _printfi + div64u)
## Finding

`gamequeer/src/bytecode.c` uses `snprintf` in two places:

1. **Line 147/150** — `SETVAR` with `TYPE_INT | TYPE_STR` flags (the `str(x)` int-to-string cast):
   ```c
   snprintf(result_str, GQ_STR_SIZE, "%ld", arg2_int);
   ```
   `%ld` pulls in `_printfi.c.obj` (2790 B) and `div64u.c.obj` (846 B — TI's `_printfi` uses 64-bit unsigned division for decimal conversion).

2. **Line 243** — `STRCAT` (the `a + b` string-concat operator):
   ```c
   snprintf(result_str, GQ_STR_SIZE, "%s%s", arg1_str, arg2_str);
   ```
   Doesn't need printf at all — bounded memcpy + memcpy.

Combined: roughly **3.6 KB of FRAM-with-wait-state code** to support two operations that could be ~30 lines of straight C each.

## Location

`gamequeer/gamequeer/src/bytecode.c:147,150,243`.

## Proposed fix

- **STRCAT**: replace `snprintf("%s%s", ...)` with bounded memcpy of `arg1_str` (up to its null terminator) then `arg2_str` (clamped to remaining buffer), ensuring null-termination.
- **Int-to-str cast**: replace `snprintf("%ld", ...)` with a small dedicated signed-int-to-decimal converter. Edge cases that need test coverage: `INT_MIN` (cannot negate in two's-complement), zero, negative numbers, the 11-byte upper bound for 32-bit signed, buffer truncation at exactly `GQ_STR_SIZE-1`.

## Estimated payoff

- Removes `_printfi.c.obj` (2790 B) and `div64u.c.obj` (846 B) from the link — roughly 3.6 KB of FRAM-wait-state code.
- Faster execution of the two affected VM operations.
- Not directly per-frame — many games (e.g., `tutorial_1.gq`) don't use `str(x)` or string concat — but indirectly improves FRAM fetch bandwidth on unrelated code by shrinking `.text`.

## Context

Surfaced as suspect #5 in [`docs/perf-investigation.md`](https://github.com/duplico/qc2024/blob/default/docs/perf-investigation.md).

Recommended to write C-side unit tests for the itoa edge cases as part of this change — see the ceedling-scaffold issue for the test infrastructure.

(AI-generated via Claude Code w/ Claude Opus 4.7 (1M context))

### PR DESCRIPTION
**Pull request by duplico**
## What

Replaces the two `snprintf` call sites in `gamequeer/src/bytecode.c`:

1. **`SETVAR` int-to-str cast (`str(x)`)** — `snprintf(result_str, GQ_STR_SIZE, "%ld", ...)` → a new `static gq_itoa(t_gq_int value, char *buf, size_t buf_size)`. Uses only 32-bit unsigned arithmetic (dividing by 10 repeatedly), no 64-bit math, and negates `INT32_MIN` safely via unsigned arithmetic (`-(value + 1)) + 1u`) to avoid signed-overflow UB. Truncates and NUL-terminates identically to the `snprintf` it replaces.
2. **`STRCAT` (`+` operator)** — `snprintf(result_str, GQ_STR_SIZE, "%s%s", a, b)` → a new `static gq_str_append_bounded(dst, dst_size, &dst_pos, src, src_size)`, called twice, reproducing the same truncate-and-NUL-terminate semantics without any printf machinery.

`stdio.h` is no longer included in `bytecode.c` (nothing else in the file used it).

## Why

Per issue #263 / `docs/perf-investigation.md` (suspect #5) in `qc2024`: on the MSP430FR2476 badge link, `%ld` formatting pulls in TI's `_printfi.c.obj` (2790 B) and `div64u.c.obj` (846 B, because `_printfi`'s decimal conversion goes through 64-bit division) for what amounts to two straightforward ~30-line C routines. Removing the only two `snprintf` call sites in the VM lets the firmware link drop both objects — an estimated ~3.6 KB of FRAM-wait-state code, only confirmable with a CCS/TI-CGT build (tracked separately, see `qc2024` PR #29 which would add a CLI TI build).

This is a pure VM-runtime change — no bytecode format, struct layout, or `gqc` compiler changes. Confirmed while reading `gqc/src/gqc/datamodel.py`/`structs.py`: strings are always stored NUL-padded to the full `GQ_STR_SIZE` (21 usable chars + NUL) at rest, in cart/FRAM/heap, by both `gqc`'s codegen and this runtime's own `SETVAR`/`STRCAT` writes, so the bounded-copy replacement's assumptions match the existing storage invariant exactly.

## Validation

**Docker/cmake build** (`GQ_HEADLESS=OFF`, normal build): clean, `bytecode.c` compiles with zero warnings; the file is `clang-format --dry-run --Werror` clean against `.clang-format`.

**Host-compiled unit test** (uncommitted, in worktree only) comparing `gq_itoa`/`gq_str_append_bounded` against the removed `snprintf` calls as an oracle:
- itoa: `0`, `-1`, `1`, `INT32_MIN`, `INT32_MAX`, `-INT32_MAX`, and truncation at buffer sizes 1/2/3/5/22 — all byte-for-byte identical to `snprintf(..., "%ld", ...)`.
- strcat: empty operands, exact-fit (21 chars), 1-byte overflow, large overflow, and arg1-alone-fills-buffer — all byte-for-byte identical to `snprintf(..., "%s%s", ...)`.

**Emulator before/after (headless harness from PR #270, used for validation only, not included in this PR):** hand-authored `.gq` game exercising `str(12345)` (literal-arg2 path), `str(-42)` and `str(INT32_MIN)` (non-literal, via `n = -2147483647; n = n - 1;` to sidestep an unrelated `gqc` literal-parsing quirk at the INT32_MIN boundary), and a string concat that overflows the 22-byte slot (`"1234567890" + "abcdefghijklmnopqrstu"` → truncates to `"1234567890abcdefghijk"`). Built two headless binaries (unpatched vs. patched `bytecode.c`), ran both against the same compiled game for 50 ticks, and diffed the dumped framebuffers pixel-by-pixel:
- `str(12345)` and the STRCAT truncation case: **pixel-identical** before/after.
- `str(-42)` and `str(INT32_MIN)`: **differ** before/after — see note below.

Pixel diff was confined exactly to the two label rows for the negative-int cases; nothing else on the 127×127 framebuffer changed.

### Bonus fix (unrelated, discovered during validation)

The pre-existing `snprintf(result_str, GQ_STR_SIZE, "%ld", arg2_int)` for negative values was already broken specifically on this LP64 Linux emulator build: `%ld` expects an 8-byte `long` on this host, but `arg2_int` is a 4-byte `int32_t` — a varargs width mismatch (UB). Before this change, `str(-42)` rendered as `4294967254` and `str(INT32_MIN)` rendered as `2147483648` in the emulator (confirmed with a standalone repro reproducing the exact same corrupted values outside the VM). MSP430 CGT's `long` is 32 bits, so the actual firmware target was never affected by this — but this change also fixes it for the Linux emulator as a side effect, since `gq_itoa` only ever operates on genuine `t_gq_int` (int32_t) values.

## Lockstep

No bytecode/struct/`gqc` changes — this is purely internal to `run_code()`'s handling of the existing `SETVAR` (int→str cast) and `STRCAT` opcodes. Confirmed by reading `gqc/src/gqc/commands.py`/`datamodel.py`/`structs.py`.

Closes #263.

(AI-generated via Claude Code w/ Claude Sonnet 5)

## Follow-up

A regression test for `gq_itoa`/`gq_str_append_bounded` (covering the itoa oracle sweep, `INT32_MIN` handling, and STRCAT truncation/overflow cases exercised manually during review) should be added to the repo's test suite as soon as the ceedling unit-test scaffold (issue #266) lands. The adversarial review pass on this PR found this logic to be correct via exhaustive manual/host validation, but that verification currently has no repo-visible, re-runnable test protecting it from regressions.

*(Edit AI-generated via Claude Code w/ Claude Fable 5)*

### CODE REVIEW

File: gamequeer/src/bytecode.c [hunk 1]
```diff
@@ -1,9 +1,71 @@
...
+#include <stddef.h>
 
 #include "HAL.h"
 #include "gamequeer.h"
 #include "gamequeer_bytecode.h"
 
+// Converts a signed 32-bit integer to a NUL-terminated decimal string,
+// writing at most buf_size - 1 characters plus the terminator (matching the
+// truncation semantics of `snprintf(buf, buf_size, "%ld", value)`, which
+// this replaces). Avoids pulling in TI's _printfi/div64u for the badge link.
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/bytecode.c]: Valid — reworded. `%ld` with an `int32_t` argument is itself a varargs width mismatch (UB) on LP64 hosts, so it wasn't a well-defined reference semantics to cite. Fixed in 9daaecb: the comment now references a correctly 32-bit-widened format specifier instead.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/bytecode.c [hunk 2]
```diff
@@ -1,9 +1,71 @@
...
+    if (negative && pos < buf_size - 1) {
+        buf[pos++] = '-';
+    }
+
+    while (ndigits > 0 && pos < buf_size - 1) {
+        buf[pos++] = digits[--ndigits];
+    }
+
+    buf[pos] = '\0';
+}
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/bytecode.c:60]: Valid — real gap, though pre-existing behavior carried over from the old snprintf-based code (it never zeroed the tail either). Fixed in 9daaecb: added a gq_str_zero_pad helper and called it after the terminator in gq_itoa, so the full GQ_STR_SIZE buffer copied to VM memory via gq_memcpy_from_ram no longer carries stale stack bytes. Same pattern already exists in menu.c's trim_menu_text_result() for precedent. Verified with a standalone ASan/UBSan harness across value/buffer-size cases (poisoned buffer, confirmed tail bytes now zero).

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/bytecode.c [hunk 3]
```diff
@@ -236,13 +298,17 @@ void run_code(t_gq_pointer code_ptr) {
...
+            case GQ_OP_STRCAT: {
                 // TODO: Break out into a function, maybe?
+                size_t result_len = 0;
                 gq_memcpy_to_ram((uint8_t *) arg1_str, cmd.arg1, GQ_STR_SIZE);
                 gq_memcpy_to_ram((uint8_t *) arg2_str, cmd.arg2, GQ_STR_SIZE);
-                snprintf(result_str, GQ_STR_SIZE, "%s%s", arg1_str, arg2_str);
+                gq_str_append_bounded(result_str, GQ_STR_SIZE, &result_len, arg1_str, GQ_STR_SIZE);
+                gq_str_append_bounded(result_str, GQ_STR_SIZE, &result_len, arg2_str, GQ_STR_SIZE);
+                result_str[result_len] = '\0';
                 gq_memcpy_from_ram(cmd.arg1, (uint8_t *) result_str, GQ_STR_SIZE);
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/bytecode.c:322]: Valid — same class of issue as the gq_itoa comment (also pre-existing under the old snprintf-based STRCAT code). Fixed in 9daaecb: call the new gq_str_zero_pad helper after result_str[result_len] = '\0' so the tail is zeroed before the GQ_STR_SIZE-byte copy to VM memory. Verified with a standalone ASan/UBSan harness (poisoned result_str, several overflow/empty-string cases) confirming truncation semantics match the prior snprintf("%s%s") behavior and the tail is now clean.

(AI-generated via Claude Code w/ Fable 5)
